### PR TITLE
Working subset of #2337 (OpenSSL 3.0)

### PR DESCRIPTION
**Pull request by Jakuje**
More minor changes that should work and be ready for merging for OpenSSL 3.0 support. Subset of larger PR #2337, which still needs some work.

##### Checklist

- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [ ] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

**Comment by Jakuje** [comment]
any comments here @dengert or @frankmorgner ? I would like to include this in the 0.22.0 so it is possible to build it with coming OpenSSL 3.0 (yet with deprecated warnings).

**Comment by dengert** [comment]
Its a good start. But I have not tried it. 

I was hoping all the SM routines that use of 3DES were also able to use AES. @frankmorgner wanted me to convert the PIV  SM code #2053 to use the existing SM code, but all the SM code appears to use 3DES, where the PIV code uses AES.

**Comment by Jakuje** [comment]
Looks like that. But even if it could save some code lines in the piv driver for creating new API for AES SM, it would be worth it, I think.

**Review by frankmorgner — COMMENTED** [review]
This looks good. The only meaningful changes for the 3des functions are covered in unit tests, so we can be quite confident that it works OK. Thank you.

**Comment by frankmorgner** [comment]
Regarding SM, I was requesting that the ISO-SM **encoding** API should be used which is given in `src/sm/sm-iso.h`. This API requires you to implement the callbacks `encrypt`, `decrypt` and `authenticate` to implement whatever crypto you need. More specifically, in `src/sm/sm-eac.c` these call backs are implemented by calling `EAC_decrypt/encrypt/authenticate` which effectively use AES256 (cmac)... If you need more help, let's use  #2053 instead.

File: src/libopensc/sc-ossl-compat.h [hunk 1]
```diff
@@ -273,6 +273,16 @@ static sc_ossl_inline void CRYPTO_secure_malloc_done()
 
 #endif /* OPENSSL_VERSION_NUMBER < 0x10100000L */
 
+/* OpenSSL 3.0 changes return value of EVP_PKEY_get0_*() to const */
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/libopensc/sc-ossl-compat.h]: Thank you for suggestion. Let me try that with 3.0.

[the remark this answers is not in the document]
Comment (Jakuje) [reply src/libopensc/sc-ossl-compat.h]: This looks like better suggestion. Tested just now with alpha15 and all things work as expected:
```
export CFLAGS="-Wno-deprecated-declarations"
./bootstrap && ./configure && make -j4
```

**Comment by Jakuje** [comment]
Just pushed updated version (with couple of minor stuff I had in working directory and that I considered too small to create a separate PR for)

**Review by frankmorgner — APPROVED** [review]

**Review by frankmorgner — APPROVED** [review]

**Comment by dengert** [comment]
p11ttest does not compile against OpenSSL-3.0.0 when OpenSSL is built in non standard locations. 

Makefile.am will use (OPTIONAL_OPENSSL_LIBS but compiles against the openssl header file from /usr/include
One line needs to be changed in src/tests/p11test/Makefile.am

```
diff --git a/src/tests/p11test/Makefile.am b/src/tests/p11test/Makefile.am
index abc2d5a1..010d710b 100644
--- a/src/tests/p11test/Makefile.am
+++ b/src/tests/p11test/Makefile.am
@@ -31,7 +31,7 @@ p11test_SOURCES = p11test.c p11test_loader.c \
        p11test_case_pss_oaep.c \
        p11test_case_interface.c \
        p11test_helpers.c
-p11test_CFLAGS = $(CMOCKA_CFLAGS)
+p11test_CFLAGS = $(OPTIONAL_OPENSSL_CFLAGS) $(CMOCKA_CFLAGS)
 p11test_LDADD = $(OPTIONAL_OPENSSL_LIBS) $(CMOCKA_LIBS)
 
 if WIN32
```

In my case OpenSSL 3.0.0 is in /opt/ossl-3.0.0/

**Comment by Jakuje** [comment]
@dengert thank you for testing. I had openssl in standard location so I did not hit it. Please, feel, free to submit a PR for this or push it to master. I think it is worth to have it in the next release.

**Comment by dengert** [comment]
Pushed aebebac432341f33d95f805ca2708fca1636e5fc to master to fix p11test/Makefile.am

**Pull request merged**